apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryInsertRow.cs

```diff
@@ -197,54 +197,42 @@ internal TableDataInsertAllRequest.RowsData ToRowsData(bool allowEmptyInsertIds)
+        private static object ConvertRowValue(object value) => value switch
         {
-            // Types that require special handling
-            switch (value)
+            null => null,
+            DateTime dt => dt.Kind switch
             {
-                case null:
-                    return null;
-                case DateTime dt:
-                    // TODO: three patterns in C# 8 with property matching?
-                    switch (dt.Kind)
-                    {
-                        // Civil datetime
-                        case DateTimeKind.Unspecified:
-                            return dt.ToString("yyyy-MM-dd'T'HH:mm:ss.FFFFFF", CultureInfo.InvariantCulture);
-                        // Timestamp
-                        case DateTimeKind.Utc:
-                            return dt.ToString("yyyy-MM-dd'T'HH:mm:ss.FFFFFF'Z'", CultureInfo.InvariantCulture);
-                        // We can get into this situation if we accept a list of DateTime values. We can't
-                        // validate when we're given the list, as it can change.
-                        default:
-                            throw new InvalidOperationException("Local DateTime values are not supported");
-                    }
-                case DateTimeOffset dto:
-                    // Timestamp
-                    return dto.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss.FFFFFF'Z'", CultureInfo.InvariantCulture);
-                case TimeSpan ts:
-                    if (ts < TimeSpan.Zero || ts >= TimeSpan.FromHours(24))
-                    {
-                        throw new InvalidOperationException("TimeSpan values must be non-negative and less than 24 hours");
-                    }
-                    // This is ugly, but it avoids a trailing . when there are no fractional seconds.
-                    return (new DateTime(1970, 1, 1) + ts).ToString("HH:mm:ss.FFFFFF", CultureInfo.InvariantCulture);
-                case BigQueryNumeric numeric:
-                    return numeric.ToString();
-                case BigQueryBigNumeric bigNumeric:
-                    return bigNumeric.ToString();
-                case BigQueryGeography geography:
-                    return geography.ToString();
-                case BigQueryInsertRow row:
-                    return row.GetJsonValues();
-            }
+                // Civil datetime
+                DateTimeKind.Unspecified => dt.ToString("yyyy-MM-dd'T'HH:mm:ss.FFFFFF", CultureInfo.InvariantCulture),
+                // Timestamp
+                DateTimeKind.Utc => dt.ToString("yyyy-MM-dd'T'HH:mm:ss.FFFFFF'Z'", CultureInfo.InvariantCulture),
+                // We can get into this situation if we accept a list of DateTime values. We can't
+                // validate when we're given the list, as it can change.
+                _ => throw new InvalidOperationException("Local DateTime values are not supported"),
+            },
```

Comment:
We can do this with multiple patterns:

```csharp
DateTime dt { Kind: Unspecified } => ...,
DateTime dt { Kind: Utc } => ...,
DateTime _ => throw ...
```

Likewise for `TimeSpan`

Reply:
For TimeSpan it might be simpler to have a private static validation method actually, such as:

```csharp
private static TimeSpan ConvertTimeSpanToDateTime(TimeSpan value)
{
    // if statement as per original code
    return new DateTime(1970, 1, 1) + value;
}
```

Then the pattern can just have:

```csharp
// Note: we convert to a DateTime as the formatting is simpler. This avoids a trailing . when there are no fractional seconds.
TimeSpan ts => ConvertTimeSpanToDateTime(ts).ToString("HH:mm:ss.FFFFFF", CultureInfo.InvariantCulture),
```

Reply:
Thanks @jskeet. This is much neater :) Done.